secretsstoresynccontroller/api/v1alpha1/secretstoresync_types.go

```diff
@@ -0,0 +1,235 @@
+	// sourcePath is the data source value of the secret defined in the Secret Provider Class.
+	// This matches the path of a file in the MountResponse returned from the provider.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=^[A-Za-z0-9.]([-A-Za-z0-9]+([-._a-zA-Z0-9]?[A-Za-z0-9])*)?(\/([0-9]+))*$
+	// +kubebuilder:validation:Required
+	SourcePath string `json:"sourcePath"`
+
+	// targetKey is the key in the Kubernetes secret's data field as described in the Kubernetes API reference:
+	// https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/secret-v1/
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=^[A-Za-z0-9.]([-A-Za-z0-9]+([-._a-zA-Z0-9]?[A-Za-z0-9])*)?(\/([0-9]+))*$
+	// +kubebuilder:validation:Required
+	TargetKey string `json:"targetKey"`
+}
+
+// SecretObject defines the desired state of synchronized Kubernetes secret objects.
+type SecretObject struct {
+	// type specifies the type of the Kubernetes secret object,
+	// e.g. "Opaque";"kubernetes.io/basic-auth";"kubernetes.io/ssh-auth";"kubernetes.io/tls"
+	// The controller must have permission to create secrets of the specified type.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Required
+	Type string `json:"type"`
+
+	// data is a slice of SecretObjectData containing secret data source from the Secret Provider Class and the
+	// corresponding data field key used in the Kubernetes secret object.
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:Required
+	// +listType=map
+	// +listMapKey=targetKey
+	Data []SecretObjectData `json:"data"`
+
+	// labels contains key-value pairs representing labels associated with the Kubernetes secret object.
+	// The following label prefix is reserved: secrets-store.sync.x-k8s.io/.
+	// The labels are used to identify the secret object created by the controller.
+	// On secret creation, the following label is added: secrets-store.sync.x-k8s.io/secretsync=<secret-sync-name>.
+	// Creation fails if the label is specified in the SecretSync object with a different value.
+	// On secret update, if the validation admission policy is set, the controller will check if the label
+	// secrets-store.sync.x-k8s.io/secretsync=<secret-sync-name> is present. If the label is not present,
+	// controller fails to update the secret.
+	// +kubebuilder:validation:XValidation:message="Labels should have < 63 characters for both keys and values.",rule="(self.all(x, x.size() < 63 && self[x].size() < 63) == true)"
+	// +kubebuilder:validation:XValidation:message="Labels should not contain secrets-store.sync.x-k8s.io. This key is reserved for the controller.",rule="(self.all(x, x.contains('secrets-store.sync.x-k8s.io') == false))"
+	// +optional
+	Labels map[string]string `json:"labels,omitempty"`
```

Comment:
@jpbetz @cici37 we ran into a gap here where CEL validation of the key hits budget limits because there's no way to constrain the maxLength of the key in openapi, so the worst-case 1-3MB size is used for the key. This meant a normal regex to check qualified name got rejected as too expensive.

Reply:
Probably worth opening an issue for kubebuilder to be able to specify constraints on the value (like maxLength). CRDs can express this but it doesn't seem like kubebuilder can.